### DISCUSSION TITLE
[ZK Stack] - Bug on `block.timestamps`

### DISCUSSION THREAD

**Discussion by thomas-nguy** [zkSync-Community-Hub/zksync-developers#820] [ZK Stack] [answered]
### Team or Project

Cronos zKEVM

### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [X] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.82

### Operating System

- [ ] Ubuntu
- [X] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

We received complains from developers that the block.timestamp is not accurate.

The block.timestamp refers to the L2 block timestamps and it can be much smaller than the actual time. For example, if we send a tx at 04:30, the block.timestamp could be 04:25 which is 5 minutes before the actual time.

This can be easily reproduced on the zkEVM testnet when the chain is idle. If there is no new block generate for few minutes, then you send a tx, the new block will have a timestamp which is few minutes earlier than the actual time.

Looking at the code, I notice that we are not allowed to produce empty L2 blocks

```
    fn should_seal_l2_block(&mut self, manager: &UpdatesManager) -> bool {
        !manager.l2_block.executed_transactions.is_empty()
            && millis_since(manager.l2_block.timestamp) > self.l2_block_commit_deadline_ms
    }
```    

Imo an easy fix would be to allow empty blocks so that block.timestamp can have a better reflection of real time

Alternatively we can also the l2 block creation logic to start the creation only when the first transaction come in. 

Anyways looking for suggestion?

### Expected Behavior

block.timestamp return real time

### Current Behavior

block.timestamp return l2 block timestamp which is not an indication of real time

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
Thank you for providing detailed information about the issue with `block.timestamp` on the ZKsync Era server module. The behavior you described, where the timestamp of the L2 block is not accurately reflecting the real time, is indeed a concern for accurate transaction processing and smart contract execution.

Your suggestion to allow empty blocks to be produced to better reflect real-time in `block.timestamp` is a valid approach. This would ensure that the timestamp is updated regularly, even when there are no transactions, thus maintaining a closer alignment with the actual time.

I recommend discussing this potential solution further with the development team or contributing a proposal for this change. You can engage with the ZKsync development community for more in-depth discussions and feedback on this issue at the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

If you have any more questions or need further assistance, feel free to ask!

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by dutterbutter** [comment]
Hey @thomas-nguy 👋 

Thanks for posting! Let me look into this for you.

**Comment by shahar4** [comment, marked as answer]
Thanks for the fantastic suggestion @thomas-nguy! As discussed, this should be implemented by creating the miniblock only when there's a tx to execute, similar to batch creation logic. I'm resolving this issue while waiting for the PR. Thanks!
